### PR TITLE
Task to generate Painless API's per context

### PR DESCRIPTION
**Pull request by jdconrad**
This adds a gradle task called generateContextDoc in the Painless module.  The task will start a cluster, issue commands against the context rest api for Painless, and generate documentation for each API per context.  Each context has a first page of classes sorted by package first and class name second, along with a page per package with each classes' constructors, methods, and fields.  A link is generated for each constructor, method, and field to a JavaDoc page when possible.

The change itself is small, but the number of lines is large due to the auto-generated changes in documentation.

Follow up: 1) Internal documentation. 2) Appropriately link each context to its API docs. 3) Add static methods/bindings.  4) Add class inheritance hierarchy.  5) Links per class/package.  6) Improve method links for declaring classes.  7) Explore reduction of total classes using hierachy.  8) Open-ended exploration of more automation for documentation per context.

Concern: With all the new links, the docs build process has slowed down some.  I really hope that I don't have to reduce the number because I think the layout is significantly improved for figuring out what's available.

### CODE REVIEW

File: docs/painless/painless-api-reference/painless-api-reference-score/index.asciidoc
Comment (nik9000) [inline docs/painless/painless-api-reference/painless-api-reference-score/index.asciidoc]: I can't find this file in the PR. Is github's UI eating it?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java]: Its probably worth a comment about why these are public instead. It'd be nice if we didn't need it though!

File: modules/lang-painless/src/doc/java/org/elasticsearch/painless/ContextDocGenerator.java
Comment (rjernst) [inline modules/lang-painless/src/doc/java/org/elasticsearch/painless/ContextDocGenerator.java]: A minimal javadoc explaining what this tool does would be nice for future modifications and understanding by others

**Comment by nik9000** [comment]
` +246,126 −6,464 ` Ah hahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahahaha.

Wow.

**Review by nik9000 — COMMENTED** [review]
Ah! I see the nature of the combinatorial explosion we're dealing with here. Do you think it'd be kinder to make files with the extra APIs that each context add on top of the global context? That'd save a ton of bits, speed up the docs build, and make it super clear what part of each context is unique.

**Review by rjernst — COMMENTED** [review]
This looks ok, but I think we should see how much work deduplication is. With this change every context gets its own copy of every whitelisted class/method, while the vast majority of these are exactly the same.

**Comment by jdconrad** [comment]
@nik9000 and @rjernst Thanks for the initial reviews.  Agree completely on consolidating the APIs where possible.  Going to add a WIP to this review for now, and work on that for the initial check in because it's clearly a bit of an increase in lines ;)

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy
Comment (jdconrad) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: Understood.  I'm confident @rjernst can take a look at those :)

File: docs/painless/painless-api-reference/painless-api-reference-score/index.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-score/index.asciidoc]: They're added now :).  Forgot to add the new files here.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java]: Would you prefer these be getters instead or comments here?

File: modules/lang-painless/src/doc/java/org/elasticsearch/painless/ContextDocGenerator.java
Comment (jdconrad) [reply modules/lang-painless/src/doc/java/org/elasticsearch/painless/ContextDocGenerator.java]: Agreed.  Will add.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java
Comment (nik9000) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextClassBindingInfo.java]: Hmmmm. I think getters would be more "normal java" to be honest. Personally, I think `public final` is fine but I think we're trying to be more "normal java".... :man_shrugging:

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy
Comment (mark-vieira) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: What's the reason for this specific change?

File: modules/lang-painless/build.gradle
Comment (mark-vieira) [inline modules/lang-painless/build.gradle]: @atorok we should probably avoid using `ClusterFormationTasks` and instead use the `TestClustersPlugin` here if that's possible. Would that work here?

File: modules/lang-painless/build.gradle
Comment (mark-vieira) [inline modules/lang-painless/build.gradle]: If we could add proper input/outputs to this task so we could benefit from UP-TO-DATE checking and eventually the build cache that would be great.

https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_input_output_runtime_api

File: modules/lang-painless/build.gradle
Comment (mark-vieira) [inline modules/lang-painless/build.gradle]: This section is redundant actually. The default convention is for source to live in `src/<<source set name>>/java`.

**Review by mark-vieira — DISMISSED** [review]

File: modules/lang-painless/build.gradle
Comment (rjernst) [reply modules/lang-painless/build.gradle]: We tried using testclusters, but it there are issues with it here. I don't think it is ready for general use yet.

Comment (atorok) [reply modules/lang-painless/build.gradle]: We are using it for all plugins  and modules now, so i'm +1 on adding new tests with testclusers. Wouldn't determinedly be interested in hearing why it didn't work.

File: modules/lang-painless/build.gradle
Comment (jdconrad) [reply modules/lang-painless/build.gradle]: Fixed.

File: modules/lang-painless/build.gradle
Comment (jdconrad) [reply modules/lang-painless/build.gradle]: I would prefer to defer this to a different PR for now.  This task is run separately from the standard build and is normally not be run in general. The infrastructure to effectively check the inputs (new script contexts, changes to resource file whitelists, and changes to the code generation) isn't fully supported, yet.  Once this task is integrated as part of the standard build, I would be happy to re-address this.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy
Comment (jdconrad) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy]: Removed as it's not needed.  Good catch.  Believe this was left behind when figuring out why the task wasn't working.

File: modules/lang-painless/build.gradle
Comment (mark-vieira) [reply modules/lang-painless/build.gradle]: I agree with @atorok we are full swing migrating at this point. We should have very good reason for _adding_ new usages of `ClusterFormationTasks`. The only valid reason I could see is if test clusters does not yet support a specific capability required by these tests.

File: docs/painless/painless-contexts-api-reference.asciidoc
Comment (jdconrad) [reply docs/painless/painless-contexts-api-reference.asciidoc]: @nik9000 If i leave it as painless-api-reference (need to change the name of the asciidoc) will this negate the need for a redirect as saved browser links will go to the new page at the current url?

File: modules/lang-painless/build.gradle
Comment (rjernst) [reply modules/lang-painless/build.gradle]: IIRC, testclusters didn't work before because painless is a module, and even though we were not using it for integTest, the check in PluginBuildPlugin would throw an error since the testclusters plugin was applied in a module. Since it is now used by modules, we may be able to get it working.

Comment (mark-vieira) [reply modules/lang-painless/build.gradle]: Yeah. I think that is supported now. @atorok can confirm.

Comment (rjernst) [reply modules/lang-painless/build.gradle]: Testclusters still does not work here. I think there needs to be more buildSrc tests for standalone uses, as I see a lot of initialization and wiring done inside PluginBuildPlugin. The configuration we tried was:

```
testClusters {
  docExtractor {
    version = project.version
    javaHome = project.file(project.ext.runtimeJavaHome)
    distribution = "DEFAULT"
  }
}

task generateContextDoc(type: JavaExec) {
  main = 'org.elasticsearch.painless.ContextDocGenerator'
  classpath = sourceSets.doc.runtimeClasspath
  useCluster testClusters.docExtractor
  systemProperty "cluster.uri", "${-> testClusters.docsExtractor.httpSocketURI}"
}
```

In addition to the extra settings of version and java home that should be defaulted to reasonable values and not require setting explicitly,  it seems to fail to depend on actually building the distribution:
> Caused by: org.elasticsearch.gradle.testclusters.TestClustersException: Can not start node{:modules:lang-painless:docExtractor-1}, missing: /mnt/ramdisk/elasticsearch/build/testclusters/extract/org.elasticsearch.distribution.elasticsearch/elasticsearch-8.0.0-SNAPSHOT

I don't think this PR should be blocked on making test clusters work for non integ tests.

File: docs/painless/painless-contexts-api-reference.asciidoc
Comment (nik9000) [reply docs/painless/painless-contexts-api-reference.asciidoc]: It will. I'm fine with that. Should the file's name stay the same?

Comment (jdconrad) [reply docs/painless/painless-contexts-api-reference.asciidoc]: Changed it back.  Good catch, ty!

**Comment by jdconrad** [comment]
@nik9000 and @rjernst Thanks for the reviews.  I believe I have addressed all the feedback appropriately and this is ready for another round.

@mark-vieira Are you okay with the gradle task as it is for now based on the feedback from @rjernst ?  As soon as the compatibility issues are handled, I'm happy to switch the task to use test clusters.

**Comment by mark-vieira** [comment]
> @mark-vieira Thanks for the review. Are you okay with the gradle task as it is for now based on the feedback from @rjernst ?

That's fine. We can revisit this as part of the ongoing effort to migrate to test clusters.

File: docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc
Comment (nik9000) [inline docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc]: I think we should cut this out of the output. It isn't really "public".

File: docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc
Comment (nik9000) [inline docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc]: Should there be more stuff here?

File: docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc
Comment (nik9000) [inline docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc]: I wonder if we shouldn't emit empty files. It isn't a big deal but I figure it'd be nice.

File: docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc
Comment (nik9000) [inline docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc]: I think we should have a note in any generated files. I havne't checked if this is generated though. If it isn't generated, I think it probably should be.

File: docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc]: I agree with this sentiment.  I don't have a great idea for how to do this and keep it automated.  Any suggestions?

File: docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc]: Agreed, and I've thought of this issue too.  I would prefer to add this in a follow up PR as this one is already getting long.

File: docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc]: It's generated.  Will add a note.

File: docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc]: This is a direct translation from the existing whitelist.  I do want to make it more sophisticated so that it can recognize things that may not need to be included, but that I think that's probably a different PR.

File: docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc]: For now I've added a hardcoded list that can be added to easily in the meantime.  In the future I would like to make this part of the entire whitelisting process where we can mark a class as internal and have it included in the PainlessLookup.  Possibly, we can even ensure that a user can't execute the method with an internal versus external compilation flag.

File: docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc
Comment (nik9000) [reply docs/painless/painless-api-reference/painless-api-reference-field/packages.asciidoc]: :+1:

File: docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc
Comment (nik9000) [reply docs/painless/painless-api-reference/painless-api-reference-aggs/packages.asciidoc]: :+1:

File: docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc
Comment (nik9000) [reply docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc]: :+1:

File: docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc
Comment (nik9000) [reply docs/painless/painless-api-reference/painless-api-reference-interval/index.asciidoc]: :+1:

File: docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc
Comment (jdconrad) [reply docs/painless/painless-api-reference/painless-api-reference-interval/packages.asciidoc]: Decided to fit this one in and fix it.

**Review by nik9000 — APPROVED** [review]
LGTM

I noticed in the "score" example it said `The standard Painless API is available.` and had a link to the root of the API page. This should probably mention that there is a special score API and maybe link directly to the standard API and the score API rather than the root, but I think that is a "for later" thing.

LGTM

**Comment by jdconrad** [comment]
@nik9000 Thanks for the multiple rounds of reviews! You're right about the need for changes to the contexts documentation at this point.  I'll follow up in another PR with changes for those.

@mark-vieira Are you okay with this code being committed as is?  I didn't want to dismiss your review requesting changes without asking.

**Pull request merged**
